CB-9180 Freeipa install in GCP does not populate reverse DNS entries

cegganesh84 opened the pull request:
1. We accept the subnet name from the user for GCP.
2. When we fetch the subnet information back from GCP, it gives the internal subnet id.
3. So make the subnet filtering logic in NetworkService similar to CloudNetworkService https://github.com/hortonworks/cloudbreak/blob/master/environment/src/main/java/com/sequenceiq/environment/network/CloudNetworkService.java#L130
4. We saw that the jinja2 template picked the next argument as the reverse zone value if the zones is an empty list, fixed that part also.

Tested in private stack

```
    $ ipa dnszone-find
      Zone name: 114.10.in-addr.arpa.
      Active zone: TRUE
      Authoritative nameserver: ipaserver0.demo5.r6zv-kv2u.dev.cldr.work.
      Administrator e-mail address: hostmaster.demo5.r6zv-kv2u.dev.cldr.work.
      SOA serial: 1602118320
      SOA refresh: 3600
      SOA retry: 900
      SOA expire: 1209600
      SOA minimum: 3600
      Allow query: any;
      Allow transfer: none;
    
      Zone name: demo5.r6zv-kv2u.dev.cldr.work.
      Active zone: TRUE
      Authoritative nameserver: ipaserver0.demo5.r6zv-kv2u.dev.cldr.work.
      Administrator e-mail address: hostmaster.demo5.r6zv-kv2u.dev.cldr.work.
      SOA serial: 1602118320
      SOA refresh: 3600
      SOA retry: 900
      SOA expire: 1209600
      SOA minimum: 3600
      Allow query: any;
      Allow transfer: none;
    ----------------------------
    Number of entries returned 2
    ----------------------------
```

lacikaaa reviewed (changes requested):
while making the salt state more error proof is good thing, I think we should add a validation to the java code, so it would fail with a specific error. It doesn't make sense to start the installation without reverse zones.
Could you add this also? Somewhere in `FreeIpaConfigService` would be a good candidate

cegganesh84 commented:
> while making the salt state more error proof is good thing, I think we should add a validation to the java code, so it would fail with a specific error. It doesn't make sense to start the installation without reverse zones.
> Could you add this also? Somewhere in `FreeIpaConfigService` would be a good candidate

Thanks for the feedback. I would certainly love to do more and more defensive programming. But what I have really done here is being safe while doing so. I just helped to fix the problem that would have happened anyway. Not want to enforce it for every other cloud provider including ycloud and future cloud providers.

I will give one example, the reverse DNS will work within the cluster. The unbound is indeed taking care of that. The only thing that breaks is idbroker requests from datahub.

```
$ cat /etc/unbound/conf.d/00-cluster.conf 
server:
  local-data: "gk2gcp-idbroker0.demo6.r6zv-kv2u.dev.cldr.work A 10.114.16.30"
  local-data-ptr: "10.114.16.30 gk2gcp-idbroker0.demo6.r6zv-kv2u.dev.cldr.work"
  local-data: "gk2gcp-master0.demo6.r6zv-kv2u.dev.cldr.work A 10.114.16.31"
  local-data-ptr: "10.114.16.31 gk2gcp-master0.demo6.r6zv-kv2u.dev.cldr.work"
  local-data: "cluster-manager.demo6.r6zv-kv2u.dev.cldr.work A 10.114.16.31"
```

I know you would still want that change! What we can negotiate is, this change will be done later, not in this commit. I am afraid it will break something and the whole change will be reverted after 1 month, putting the GCP plan in jeopardy.

cegganesh84 commented on freeipa/src/main/java/com/sequenceiq/freeipa/service/stack/NetworkService.java:
```diff
@@ -57,11 +57,13 @@
                     String[] splittedNetworkId = cloudNetwork.getId().split("/");
                     String cloudNetworkId = splittedNetworkId[splittedNetworkId.length - 1];
-                    return networkId.equals(cloudNetworkId);
+                    // CB-9180 For GCP user can pass network name as id
```
Actually, I have a different opinion here. It is interesting that both are valid points from each perspective.

A person who is reading that line of code, and click blame to read through my commit message. But a casual reader will miss that.

True story, `// support for azure` is actually what made me stop and think about what might be going wrong in that function for GCP!

lacikaaa replied:
I have to say I would do it differently today, move that part to a method and give a meaningful name. If we add these comments to every part of the code then it will be outdated in a while etc. I still prefer avoiding comments as much as possible. Solving that issue and indicating with a comment was a mistake from my part.

cegganesh84 replied:
Ok, I have removed those comments to move forward.

lacikaaa commented:
my point there you can't break anything as it worked everywhere for every cloud provider until now, because there was no way there could be any empty string/null value in `String reverseZones`.
If we leave it this way, it would be harder to identify later the problem as there won't be suspicious reverse zones etc. I'm fine with not including the validation in this PR but I would like to see a jira or PR addressing that issue, so we won't miss that.

Also unbound has reverse pointers intracluster only, so that's not sufficient. And this could cause issues for experiences too.

Please squash your commits into a single one.

cegganesh84 commented:
> my point there you can't break anything as it worked everywhere for every cloud provider until now, because there was no way there could be any empty string/null value in String reverseZones.

I am not sure if that holds true always. What if a cloud provider like ycloud could have given an ability to auto-reverse all the hostnames?

 > Please squash your commits into a single one.

I will, during the merge I guess GitHub supports squash and merge.

cegganesh84 commented:
> If we leave it this way, it would be harder to identify later the problem as there won't be suspicious reverse zones etc. I'm fine with not including the validation in this PR but I would like to see a jira or PR addressing that issue, so we won't miss that.
> 
> Also unbound has reverse pointers intracluster only, so that's not sufficient. And this could cause issues for experiences too.
> 
> Please squash your commits into a single one.

I have added a comment to the jira that it will remain open until this PR comment is addressed https://jira.cloudera.com/browse/CB-9180?focusedCommentId=4084957&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-4084957

Removed the comments and squashed my commits.

cegganesh84 commented:
@lacikaaa thanks for the approval. I can not merge it, please merge it, it does not have any conflicts.
